lib/icalendar/rrule.ex

```diff
@@ -0,0 +1,420 @@
+  """
+
+  alias ICalendar.Property
+  alias ICalendar.Util.Deserialize
+
+  @frequencies %{
+    "SECONDLY" => :secondly,
+    "MINUTELY" => :minutely,
+    "HOURLY"   => :hourly,
+    "DAILY"    => :daily,
+    "WEEKLY"   => :weekly,
+    "MONTHLY"  => :monthly,
+    "YEARLY"   => :yearly
+  }
+
+  @days %{
+    "SU" => :sunday,
+    "MO" => :monday,
+    "TU" => :tuesday,
+    "WE" => :wednesday,
+    "TH" => :thursday,
+    "FR" => :friday,
+    "SA" => :saturday
+  }
+
+  @months [:january, :february, :march,
+           :april, :may, :june, :july,
+           :august, :september, :october,
+           :november, :december]
+
+
+  @string_to_atom_keys %{
+    "FREQ"       => :frequency,
+    "COUNT"      => :count,
+    "UNTIL"      => :until,
+    "INTERVAL"   => :interval,
+    "BYSECOND"   => :by_second,
+    "BYMINUTE"   => :by_minute,
+    "BYHOUR"     => :by_hour,
+    "BYMONTHDAY" => :by_month_day,
+    "BYYEARDAY"  => :by_year_day,
+    "BYWEEKNO"   => :by_week_number,
+    "BYSETPOS"   => :by_set_pos,
+    "BYDAY"      => :by_day,
+    "BYMONTH"    => :by_month,
+    "WKST"       => :week_start,
+    "X-NAME"     => :x_name
+  }
+
+  defstruct frequency: nil,
+            until: nil,
+            count: nil,
+            interval: nil,
+            by_second: [],
+            by_minute: [],
+            by_hour: [],
+            by_day: [],
+            by_month: [],
+            by_month_day: [],
+            by_year_day:  [],
+            by_week_number: [],
+            by_set_pos: [],
+            week_start: nil,
+            x_name: nil,
+            errors: []
+
+  @doc ~S"""
+  Produces a list of months in the year
+  """
+  def months, do: @months
+
+  @doc ~S"""
+  Produces a list of RRULE iCal String keys and their %ICalendar.RRULE{}
+  counterparts.
+  """
+  def string_to_atom_keys, do: @string_to_atom_keys
+  def string_to_atom_keys(:inverted) do
+      @string_to_atom_keys
+      |> Enum.reduce(%{}, fn({key, value}, acc) ->
+        Map.put(acc, value, key)
+      end)
+  end
+
+  @doc ~S"""
+  Produces a list of valid RRULE frequencies and their %ICalendar.RRULE{}
+  counterparts.
+  """
+  def frequencies, do: @frequencies
+  def frequencies(:inverted) do
+    @frequencies
+    |> Enum.reduce(%{}, fn({key, value}, acc) ->
+      Map.put(acc, value, key)
+    end)
+  end
+
+  @doc ~S"""
+  Produces a list of valid RRULE days and their %ICalendar.RRULE{}
+  counterparts.
+  """
+  def days, do: @days
+  def days(:inverted) do
+    @days
+    |> Enum.reduce(%{}, fn({key, value}, acc) ->
+      Map.put(acc, value, key)
+    end)
+  end
+
+  @doc ~S"""
+  This function is used to determine whether an RRULE struct has errors or not
+
+      iex> ICalendar.RRULE.valid(%ICalendar.RRULE{errors: []})
+      true
+
+      iex> ICalendar.RRULE.valid(%ICalendar.RRULE{errors: ["error"]})
+      false
+  """
+  def valid(%ICalendar.RRULE{errors: errors}) when is_list(errors) do
+    Enum.count(errors) == 0
```

Comment:
We can pattern match on the empty list to assert that it's valid in `O(1)` rather than counting the list, which is `O(n)`